primitives/src/crh/bowe_hopwood/mod.rs

```diff
@@ -86,31 +115,10 @@ impl<G: Group, W: PedersenWindow, const N: usize> BoweHopwoodPedersenCRH<G, W, N
         Ok(())
     }
 
-    /// Zcash protocol specification - section 5.4.1.7
-    /// We need to find a c such that 4 * (2^4c - 1)/15 <= F::MODULUS_MINUS_ONE_DIV_TWO
-    pub(crate) fn calculate_num_chunks_in_segment<F: PrimeField>() -> usize {
-        let upper_limit: BigUint = (F::modulus_minus_one_div_two().into() * 15u32) + 4u32;
-
-        let mut c = 0;
-        let mut range = BigUint::from(1u32);
-        while (&range << 2) <= upper_limit {
-            range <<= 4;
-            c += 1;
-        }
-
-        // Rule out weird edge cases. Can only happen if MODULUS = 0 or 1
-        if c == 0 {
-            panic!("Invalid Field modulus (0 or 1). Unable to compute parameters.");
-        }
-
-        c - 1
-    }
-
-    pub fn create_generators<R: Rng>(rng: &mut R) -> Vec<Vec<G>> {
+    pub fn create_generators_from_base_points(base_gens: Vec<G>) -> Vec<Vec<G>> {
         let mut generators = Vec::new();
-        for _ in 0..W::NUM_WINDOWS {
+        for mut base in base_gens.into_iter().take(W::NUM_WINDOWS) {
```

Comment:
maybe let's check that base_gens.len() == W::NUM_WINDOWS instead of silently allowing it

Reply:
Ok